controller/src/controller.rs

```diff
@@ -271,27 +276,42 @@ impl<T: BottlerocketShadowClient> BrupopController<T> {
-                // controller should respect it, continue to complete the update, and stop actions on remaining nodes.
-                // This logic will cooperate with 6 mins buffer strategy.
-                false => {
-                    // try to find out if any nodes are being updated. If yes, controller
-                    // will complete them and pause actions on other waitingForUpdate nodes.
-                    if !active_set.is_empty() {
-                        self.progress_active_set(active_set).await?;
-                    } else {
-                        sleep(ACTION_INTERVAL).await;
-                        continue;
-                    }
-                }
-                true => {
-                    if !active_set.is_empty() {
-                        self.progress_active_set(active_set).await?;
-                    }
+            // complete any ongoing updates even if we are outside of a scheduled window.
+            if !active_set.is_empty() {
+                self.progress_active_set(active_set).await?;
+            } else {
+                // calculate the duration between current time and next scheduled time, and then make
+                // controller sleep until next scheduled time point.
+                let next_scheduled_time = brupop_scheduler
+                    .upcoming(Utc)
+                    .next()
+                    .context(controllerclient_error::GetScheduledDatetimeSnafu)?;
+                let duration_until_next_schedule = next_scheduled_time - Utc::now();
+
+                // covert chrono duration to std duration
+                let duration_until_next_schedule_std = duration_until_next_schedule
+                    .to_std()
+                    .context(controllerclient_error::ConvertToStdDurationSnafu)?;
+                event!(
+                    Level::INFO,
+                    next_schedule_time = ?next_scheduled_time,
+                    sleep = duration_until_next_schedule_std.as_secs(),
+                    "Began sleeping until next scheduled time point."
+                );
+                tokio::time::sleep(duration_until_next_schedule_std).await;
+            }
 
+            // Cron expression can be configured to a time window or a specific trigger time.
+            // specific trigger time: 0 0 10 * * Mon */Every Monday at 10AM.
+            // maintenance window (time window): * * 10-12 * * MON */Every Monday between 10:00 and 12:00.
+
+            // brupop controller needs to use different logics to deal with specific trigger time or maintenance window (time window)
+            // => specific trigger time: trigger brupop update and complete all waitingForUpdate nodes.
+            // => maintenance window (time window): trigger brupop update within time window. If current time isn't within the time window,
+            // controller shouldn't have any action on it.
+            if is_time_window(&brupop_scheduler)? {
```

Comment:
It looks like you mean this?

```suggestion
            if is_time_window(&brupop_scheduler)? && within_time_window(&brupop_scheduler)? {
```